core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.AlluxioURI;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.InvalidPathException;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.io.PathUtils;
+
+import io.netty.util.internal.chmv8.ConcurrentHashMapV8;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class caches paths in the UFS which do not exist.
+ */
+@ThreadSafe
+public class UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsAbsentPathCache.class);
+
+  /** The mount table. */
+  protected final MountTable mMountTable;
+  /** Stores a cache for each mount point (the key is mount id). */
+  protected final ConcurrentHashMapV8<Long, ConcurrentSkipListSet<String>> mCaches;
+
+  /**
+   * Creates a new instance of {@link UfsAbsentPathCache}.
+   *
+   * @param mountTable the mount table
+   */
+  public UfsAbsentPathCache(MountTable mountTable) {
+    mMountTable = mountTable;
+    mCaches = new ConcurrentHashMapV8<>(8, 0.95f, 8);
+  }
+
+  /**
+   * Removes the cache for the given mount id.
+   *
+   * @param mountId the mount id
+   */
+  public void removeMountPoint(long mountId) {
+    mCaches.remove(mountId);
+  }
+
+  /**
+   * Adds the given absent path into the cache. This will sequentially walk down the path to find
+   * the first component which does not exist in the ufs.
+   *
+   * @param path the absent path to add to the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void addAbsentPath(AlluxioURI path) throws InvalidPathException {
+    MountTable.Resolution resolution = mMountTable.resolve(path);
+    AlluxioURI ufsUri = resolution.getUri();
+    UnderFileSystem ufs = resolution.getUfs();
+    ConcurrentSkipListSet<String> cache = getCache(resolution.getMountId());
+
+    String[] components = PathUtils.getPathComponents(ufsUri.getPath());
+    if (components.length == 0) {
+      return;
+    }
+
+    // create a ufs uri of the root of the mount point.
+    AlluxioURI uri = mMountTable.getMountInfo(resolution.getMountId()).getUfsUri();
+    int mountPointBase = uri.getDepth();
+
+    // Traverse through the ufs path components, staring from the root, to find the first
+    // non-existing ufs path.
```

Comment:
I think that is possible, but since there is never a guarantee that alluxio is in-sync with ufs, we can just process normally. Since this will be done in a threadpool, it should not impact the critical path.